### ISSUE TITLE
Review: Affix Embedding Training

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#125]
## Overview

Review how recent changes impact affix (prefix and suffix) embedding training. This stage trains 8-dimensional embeddings for prefixes and suffixes.

### 1. Fundamento-Based Prefix Disambiguation
- **Change**: Parser now correctly identifies prefixes using Fundamento statistics
- **Potential Impact**:
  - **Prefix vocabulary is now correct**: "mal-", "re-", "ek-" properly identified
  - **False positives eliminated**: Words like "mala" (evil) no longer parsed as "mal-a"
  - **Prefix frequency distributions change dramatically**
  - **Training pairs need regeneration**: See issue #79 (affix training logic backwards)

### 2. Parser `prefiksoj` Format Change
- **Change**: `prefiksoj` is now a list (can have multiple prefixes)
- **Potential Impact**:
  - Words like "remalami" now correctly yield `["re", "mal"]`
  - Training must handle multi-prefix words
  - Embedding composition must be updated (sum? mean? ordered?)

### 3. Existing Training Issues (Still Relevant)
- **#79**: Affix training logic semantically backwards
  - Words with same prefix trained as similar, but "malbona" ≠ "malgranda"
  - Need semantic similarity signal, not just morphological similarity
- **#82**: Antonyms via mal- have no training signal
  - "bona" and "malbona" should be related but opposite
  - Current training may push them apart or together incorrectly

### 4. Corpus Tiering System
- **Change**: Fundamento sentences weighted 10x
- **Potential Impact**:
  - Fundamento uses prefixes canonically
  - "mal-" in Fundamento is always antonym, never "mala-" stem
  - Weighting will reinforce correct prefix semantics

## Questions to Answer

1. What is the complete prefix list after parser fixes?
2. What is the frequency distribution of each prefix?
3. How do we train prefixes to capture semantic transformation?
4. Should prefixes have learnable composition order (for "re-mal-")?
5. How do we handle the antonym problem with "mal-"?

## Training Artifacts to Review

- `data/vocabularies/prefix_vocab.json` - 61 prefixes
- `data/vocabularies/suffix_vocab.json` - 38 suffixes
- `scripts/train_affix_embeddings.py` - Training script (if exists)
- Issue #69 (Phase 2: Affix transformation training)

## Training Data Requirements

- **Input**: Unified corpus with correct AST parses
- **Pairs needed**: 
  - Same-root, different-affix pairs (e.g., "ami" vs "malami")
  - Semantic similarity labels (not just morphological)
- **Key insight**: Affix training should capture TRANSFORMATION, not just presence

## Dependencies

- **Blocked by**: #123 (Training Data Review), #124 (Root Embedding Review)
- **Blocks**: Sentence encoding, compositional embedding quality

**Comment by marctjones** [comment]
## Status Update (2025-12-27)

**Affix Embedding Training COMPLETE** ✅

- 12 prefixes + 31 suffixes × 32d = 1,376 params
- Trained on affix-root co-occurrence from 1.98M sentences
- 6/6 evaluation tests pass

### Validation Results
- Participial active (ant, int, ont): avg_sim=0.268 ✅
- Participial passive (at, it, ot): avg_sim=0.354 ✅
- Degree (et, eg): avg_sim=0.248 ✅
- Person/place (ul, ej): avg_sim=0.218 ✅

### Known Limitations
- 5/12 prefixes have sparse training data (dis, bo, mis, fi, vic)
- Prefix avg similarity = 0.014 (weaker than suffixes)

### Files Created
- `scripts/training/train_affix_embeddings.py`
- `scripts/training/evaluate_affix_embeddings.py`
- `data/vocabularies/affix_vocabulary.json`

This issue can be closed. See #133 for next step (refactor to enriched ASTs).

**Comment by marctjones** [comment]
## Analysis Findings (Dec 29, 2025)

After running the RAG demo with the trained affix transforms, we have **confirmed embedding collapse**:

### Similarity Matrix Evidence

```
                     bona   malbona   lerni   lernejo  lernanto   patro  patrino
        bona        1.000     0.290   0.176     0.305     0.289   0.154    0.251
        malbona     0.290     1.000   0.489     0.980     0.993   0.204    0.985
        lernejo     0.305     0.980   0.493     1.000     0.990   0.201    0.968
```

**Critical observations:**
- `bona` vs `malbona` = 0.290 (GOOD - opposites should be dissimilar)
- `malbona` vs `lernejo` = 0.980 (BAD - unrelated words with same transformation structure)
- `malbona` vs `patrino` = 0.985 (BAD - unrelated words clustering)

### Root Cause: Additive Residual Architecture

The current `LowRankTransform` (`scripts/training/train_affix_transforms.py:32-40`):
```python
class LowRankTransform(nn.Module):
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return x + self.up(self.down(x))  # ← Problem: residual addition
```

**Why this fails:**
- With small `gain=0.1` initialization, `up(down(x))` is tiny
- Result: `x + tiny_offset ≈ x` for all words
- All affixed words end up near the identity transformation

### Missing Prefixes in Training Data

Analysis shows 5 prefixes have **ZERO** training examples:
- `bo-` (in-law): 0 examples
- `dis-` (disperse): 0 examples
- `mis-` (wrongly): 0 examples
- `fi-` (morally bad): 0 examples
- `vic-` (vice-): 0 examples

These affixes cannot learn meaningful transforms with no data.

### Proposed Fix

Replace additive residual with **multiplicative transformation**:
```python
def forward(self, x):
    # Matrix form: y = W @ x where W = I + U @ V
    return x + self.up(self.down(x))  # OLD
    
    # Alternative: Rotation + scaling
    return self.transform_matrix @ x  # NEW
```

Or use **contrastive training** to push apart unrelated words.

### Related Issues
- #143 - P0: RAG producing poor results (blocked by this)
- #139 - Stage 1b: Retrain affix embeddings
